### ISSUE TITLE
Problem unit testing Time4A 3.24-2016i

### ISSUE THREAD

**Issue by pickscrape** [MenoData/Time4J#582]
I'm not entirely sure whether this has been triggered by a change in Time4A or Android Studio (maybe gradle?), but I'm now having another problem unit testing code like this:

`ChronoFormatter<PlainDate> formatter = ChronoFormatter.ofDateStyle(DisplayMode.LONG, Locale.getDefault());`

(The above code can be in the unit test itself: it doesn't need to be called in some other code).

I end up with a stack trace that looks like this:

`java.lang.ExceptionInInitializerError
	at net.time4j.format.expert.AttributeSet.<clinit>(Unknown Source)
	at net.time4j.format.expert.ChronoFormatter$Builder.startSection(Unknown Source)
	at net.time4j.format.expert.PatternType.addMonth(Unknown Source)
	at net.time4j.format.expert.PatternType.cldrISO(Unknown Source)
	at net.time4j.format.expert.PatternType.cldr(Unknown Source)
	at net.time4j.format.expert.PatternType.registerSymbol(Unknown Source)
	at net.time4j.format.expert.ChronoFormatter$Builder.addPattern(Unknown Source)
	at net.time4j.format.expert.ChronoFormatter.addPattern(Unknown Source)
	at net.time4j.format.expert.ChronoFormatter.ofDatePattern(Unknown Source)
	at net.time4j.format.expert.ChronoFormatter.ofDateStyle(Unknown Source)
	at time4atestexample.example.com.myapplication.ExampleUnitTest.itFails(ExampleUnitTest.java:21)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.util.MissingResourceException: Can't find bundle for base name numbers/symbol, locale 
	at java.util.ResourceBundle.throwMissingResourceException(ResourceBundle.java:1564)
	at java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1387)
	at java.util.ResourceBundle.getBundle(ResourceBundle.java:1299)
	at net.time4j.i18n.SymbolProviderSPI.<clinit>(Unknown Source)
	... 46 more`

It's entirely possible that I'm just not initializing the library properly, but I've tried everything I can think of (not that I know of a lot of tricks at this point) to get that working in a unit testing context, without success.

I do know that the last time I looked at this code, I didn't have this problem. So something has changed to break it, but I'm not sure what. I did just now create a completely fresh project that does nothing but add Time4A as a dependency and adds the above line to the unit test, so I know it's not something specific to the rest of my project.

Is there something I'm missing, or is this a bug?

Thanks.

**Comment by MenoData** [comment]
Thanks for feedback. I will check it in the evening. But first two questions:

What is your default locale? And are you sure you have initialized Time4A using `ApplicationStarter.initialize(this, ...)`?

**Comment by pickscrape** [comment]
Default locale **should** be en_US, but I admit that I never actually checked what that was coming back as in the unit test context. I'll check this out this evening to see if that's related.

I did try calling ApplicationStarter.initialize (with both true and false for the second parameter) via a mocked object. Getting the application object in a unit test is what I meant when I talked about "tricks": I'm still not 100% comfortable with how all of that works, so it's completely possible that I've missed some step that I should be taking here.

Thanks!

**Comment by pickscrape** [comment]
Confirmed, the locale being used is en_US.

Now, I actually have got this working, but I had to implement it as an instrumented test, which runs the test via an Android emulator:

```
Context context = InstrumentationRegistry.getTargetContext();

ApplicationStarter.initialize(context, true);

ChronoFormatter<PlainDate> formatter = ChronoFormatter.ofDateStyle(DisplayMode.LONG, Locale.getDefault());
```

That's a lot slower than just running it in the JVM, and seems like overkill for testing my date/time manipulation routines. :)

If there's a way to do without having to use instrumentation, that would be much preferable.

Thanks!

**Comment by MenoData** [comment]
Ah okay, so this problem is something about context initialization and the use of right context. First some general informations (as check list):

- The initialization via `ApplicationStarter.initialize(...)` must happen first before any call of other Time4A-code or initialization.
- The boolean-prefetch-parameter in this initializing method is not relevant for the correct initialization of resource access.
- The `MissingResourceException` you observed only happens if Time4A cannot find the own assets. In order to be capable of finding the assets, Time4A needs a suitable Android-`Context` so it can internally do resource access via `context.getAssets().open("net/time4j/i18n/numbers/symbol.properties");.

My suspicion is now that you had originally used a context which has no application context. Saying that, I have following lines of the initialization method in the class `ApplicationStarter` in mind:

```
         Context ctx = context.getApplicationContext(); 
         prepareAssets(ctx); // initialize the class AndroidResourceLoader (must be called first) 
```

Would it help to just use the context parameter directly? Namely writing:

`prepareAssets(context);` // without looking for application context

**Comment by MenoData** [comment]
I believe that the recent refactoring of class `ApplicationStarter` might fix your problem. Please be so kind to give short feedback if the next release v3.25-2016j fixes your test-setup. Since Time4A no longer insists on the application context for resource access, you have the freedom to inject any suitable context of your test environment where the application context might be absent anyway.

**Comment by pickscrape** [comment]
I'm still having trouble, unfortunately. In order to dig in a bit deeper, I created a fork (https://github.com/pickscrape/Time4A) of Time4A and committed a unit test that demonstrates what I'm trying to do.

As part of it, I made ApplicationStarter.prepareAssets public so it could be called externally as one of my attempts. None of the three methods I tried worked, though one of them does fail a little differently.

Again, I'm sure this is due to something I'm missing or getting wrong, but I'm struggling to spot it.

**Comment by MenoData** [comment]
Sorry to hear, it is good to see your fork giving a better chance that I will have a close look on your concrete test set-up and see if I can find a solution during the next days. However, I actually think that there is not really a bug in Time4A related to your problem but the last word is not said yet...

**Comment by MenoData** [comment]
Have not yet tested your fork but a first look at this code

```
20 +@RunWith(PowerMockRunner.class)  
 21 +@PrepareForTest({Log.class})  
 22 +public class InitTest  
 23 +{  
 24 +	@Mock  
 25 +	Context mContext;  
 26 +  
 27 +	@Test  
 28 +	public void testWithPrepareAssets()  
 29 +	{  
 30 +		PowerMockito.mockStatic(Log.class);  
 31 +  
 32 +		mContext = Mockito.mock(Context.class);  
 33 +  
 34 +		ApplicationStarter.prepareAssets(mContext);  
 35 +  
 36 +		PatternType patternType = PatternType.CLDR;  
 37 +		Locale locale = Locale.getDefault();  
 38 +  
 39 +		ChronoFormatter.ofDatePattern("yyyyMMdd", patternType, locale);  
 40 +	}  

```

let met doubt: How can Powermockito know where to find the assets of Time4A if it only generates an abstract `Context` in line 32 ? There is not even any reference to Time4A-classes or enviroment in that line. **You need to find a way to pass the location of assets to the context generated by PowerMockito.** 

An option might be to wrap the generated context in your own context class (using delegation in all methods) and to override the methods [getAssets()](https://developer.android.com/reference/android/content/Context.html#getAssets()) and probably `getResources()`, too. I am not familiar with Powermock but maybe Powermock offers an easy way to define the assets path (asset-manager) on the generated context.

**Comment by MenoData** [comment]
Two other points:

- Do you really need Time4A, or do you just want to test Time4J and other time-related stuff? If so then you should consider a standard JUnit-test-environment on a Java-platform (not Android).

- Otherwise, I could also introduce an interface like this in next release of Time4A:

```
public interface ResourceAccessor {
  InputStream open(String resource) throws IOException;
}
```

Then you can easily define an appropriate code base for finding the assets as soon as you can initialize Time4A with your own implementation of this interface.

**Comment by pickscrape** [comment]
Thanks for your time and feedback! I'll be digging into your suggestions later.

I definitely need Time4A because I'm specifically developing an Android app, and it's the only library I've been able to find that can localize durations. :)

It's pretty clear that the problem is that I'm not doing the right things to get things up and ready for testing. I'm going to spend some time later trying to get that working (I'm not familiar with PowerMock either: I'm feeling my way along here in a lot of ways).

Ultimately, I certainly think it should be possible to test a library such as this without having to use instrumented tests via an emulator. If that means the library needs to expose some way to get it access to its resources without needing an Application context, then that makes sense to me.

I'll let you know how I get along once I've made some more attempts.

Thanks again!

**Comment by MenoData** [comment]
I have now released Time4A-v3.28-2016j. Its class `ApplicationStarter` contains the new method:

`public static void prepareAssets(Context context, AssetLocation assetLocation)`

This enables customized access to any unusual asset environment by specialized initialization code (without calling `ApplicationStarter.initialize(...)`. You might try this based on knowledge of your project test set-up. Good luck!

**Comment by pickscrape** [comment]
Some "SOLVED!" info for anyone else who happens across this issue.

After a long hiatus from working on this, I decided to give it another try this weekend.

I tried passing the asset/resource location into the library in every way I could think of, and nothing seemed to work.

As a last resort, I decided to try importing the library into my project as a library project, rather than reference it via maven. The thinking was that this would be more likely to make the resource data available to the test process.

So instead of something like this:

compile 'net.time4j:time4j-android:3.19-2016d'

I used this:

compile project(':time4j-android')

Well, it worked! All tests now pass without having issues with resources, and I didn't have to do anything special to set up the resources either (this is possibly partially thanks to Robolectric, though I haven't tried without yet).

This will make updating the library a little more involved, but well worth the effort.

Thanks once again for your time on this one.
